File: Leetcode_POTD_2025/8_August/3363_Find_the_Maximum_Number_of_Fruits_Collected.cpp
class Solution {
public:
    int maxCollectedFruits(vector<vector<int>>& z) {
        int n = z.size(), r = 0;
        for (int i = 0; i < n; ++i) r += z[i][i];
        for (int i = 0; i < n - 2; ++i) z[i][n - 2 - i] = z[i][n - 3 - i] = 0;
        z[n - 2][0] = 0;
        for (int i = 1; i < n - 1; ++i)
            for (int j = max(i + 1, n - i - 1); j < n; ++j)
                z[i][j] += max({z[i - 1][j], z[i - 1][j - 1], j + 1 < n ? z[i - 1][j + 1] : 0});
        for (int j = 1; j < n - 1; ++j)
            for (int i = max(j + 1, n - j - 1); i < n; ++i)
                z[i][j] += max({z[i - 1][j - 1], z[i][j - 1], i + 1 < n ? z[i + 1][j - 1] : 0});
        return r + z[n - 2][n - 1] + z[n - 1][n - 2];
    }
};

File: Leetcode_POTD_2025/8_August/3363_Find_the_Maximum_Number_of_Fruits_Collected.py
class Solution(object):
    def maxCollectedFruits(self, z):
        """
        :type z: List[List[int]]
        :rtype: int
        """
        n = len(z)
        r = sum(z[i][i] for i in range(n))

        for i in range(n - 2):
            z[i][n - 2 - i] = 0
            z[i][n - 3 - i] = 0

        z[n - 2][0] = 0

        for i in range(1, n - 1):
            for j in range(max(i + 1, n - i - 1), n):
                max_val = z[i - 1][j]
                max_val = max(max_val, z[i - 1][j - 1])
                if j + 1 < n:
                    max_val = max(max_val, z[i - 1][j + 1])
                z[i][j] += max_val

        for j in range(1, n - 1):
            for i in range(max(j + 1, n - j - 1), n):
                max_val = z[i - 1][j - 1]
                max_val = max(max_val, z[i][j - 1])
                if i + 1 < n:
                    max_val = max(max_val, z[i + 1][j - 1])
                z[i][j] += max_val

        return r + z[n - 2][n - 1] + z[n - 1][n - 2]
